File: Music_Lyric/app/src/main/java/com/kuanquan/music_lyric/utils/GsonUtils.java
package com.kuanquan.music_lyric.utils;

import android.content.Context;
import android.content.res.AssetManager;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.List;
import java.util.Map;

public class GsonUtils {
    private static Gson gson = null;

    public Gson getInstance() {
        if (gson == null) gson = new GsonBuilder()
                .setPrettyPrinting()
                .disableHtmlEscaping()
                .create();
        return gson;
    }

    public static <T> T toObject(String json, Class clazz) {
        try {
            return (T) new GsonUtils().getInstance().fromJson(json, clazz);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static <T> T toObject(String json, Type type) {
        try {
            return (T) new GsonUtils().getInstance().fromJson(json, type);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String toJson(Map map) {
        return new GsonUtils().getInstance().toJson(map);
    }

    public static List<?> toList(String json, Type type) {
        Gson gson = new Gson();
        List<?> list = gson.fromJson(json, type);
        return list;
    }

    public static String toJson(Object obj) {
        Gson gson = new Gson();
        if (obj == null) {
            return "";
        } else {
            return gson.toJson(obj);
        }
    }

    public static <T> T fromJson(String str, Class<T> type) {
        Gson gson = new Gson();
        return gson.fromJson(str, type);
    }

    public static <T> T fromJsonType(String str, Type type) {
        Gson gson = new Gson();
        return gson.fromJson(str, type);
    }

    public static String getJson(Context context, String fileName) {

        StringBuilder stringBuilder = new StringBuilder();
        try {
            AssetManager assetManager = context.getAssets();
            BufferedReader bf = new BufferedReader(new InputStreamReader(
                    assetManager.open(fileName)));
            String line;
            while ((line = bf.readLine()) != null) {
                stringBuilder.append(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
//            LogUtil.e("开始解析异常");
        }
        return stringBuilder.toString();
    }
}

File: weixinEdit/baselibrary/src/main/java/com/mouble/baselibrary/util/GsonUtils.kt
package com.mouble.baselibrary.util

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import java.lang.reflect.Type

object GsonUtils {
    private var gson: Gson? = null

    fun getInstance(): Gson? {
        if (gson == null) gson = GsonBuilder()
            .setPrettyPrinting()
            .disableHtmlEscaping()
            .create()
        return gson
    }

    /**
     * 返回一个对象
     *
     * @param json
     * @param clazz
     * @param <T>
     * @return
     */
    @JvmStatic
    fun <T> toObject(json: String?, clazz: Class<*>?): T? {
        try {
            return getInstance()?.fromJson(json, clazz) as? T
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 返回一个对象
     *
     * @param json
     * @param type
     * @param <T>
     * @return
     */
    fun <T> toObject(json: String?, type: Type?): T? {
        try {
            return getInstance()?.fromJson<Any>(json, type) as? T
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 将Map参数转成json格式传输
     *
     * @param map
     * @return
     */
    fun toJson(map: Map<*, *>?): String? {
        return getInstance()?.toJson(map)
    }

    /**
     * 把json字符串变成集合
     * params: new TypeToken<List></List><yourbean>>(){}.getType(),
     *
     * @param json
     * @param type new TypeToken<List></List><yourbean>>(){}.getType() List<LeaveSchedule> tmpBacklogList = gson.fromJson(backlogJsonStr, new TypeToken<List></List><LeaveSchedule>>() {}.getType());
     * @return ArrayList<ContentMsg> json1 = gson.fromJson(s.toString(),new TypeToken<List></List><ContentMsg>>() {}.getType());
     */
    fun toList(json: String?, type: Type?): List<*>? {
//        val gson = Gson()
        return getInstance()?.fromJson<List<*>>(json, type)
    }

    /**
     * 对象转换成json字符串
     *
     * @param obj
     * @return
     */
    fun toJson(obj: Any?): String? {
        val gson = Gson()
        return if (obj == null) {
            ""
        } else {
            gson.toJson(obj)
        }
    }

    /**
     * json字符串转成对象
     *
     * @param str
     * @param type
     * @return
     */
    fun <T> fromJson(str: String?, type: Class<T>?): T {
        val gson = Gson()
        return gson.fromJson(str, type)
    }

    /**
     * json字符串转成对象或集合（数组）
     *
     * @param str
     * @param type java.lang.reflect.Type type = new TypeToken<HashMap></HashMap><Integer></Integer>, String>>() {}.getType();
     * @return
     */
    fun <T> fromJsonType(str: String?, type: Type?): T {
        val gson = Gson()
        return gson.fromJson(str, type)
    }
}